### DISCUSSION TITLE
Can I use SSD MobileNet V2 on AI Deck and how to implement object detection from scratch?

### DISCUSSION THREAD

**Discussion by ArtemRedchych** [bitcraze/discussions#640] [Q&A]
I am working with the AI Deck and trying to implement object detection using the SSD MobileNet V2 model. I have set up a Docker container with the 'aideck-with-autotiler' image, but when I try to run the nntool command, I am encountering the following error:

bash: nntool: command not found

**EDIT:**
ofter "reinstalling" the image i am getting this error when running nntool >

root@fcb3bfeebf6e:/gap_sdk# nntool
RuntimeError: module compiled against API version 0xf but this version of numpy is 0xd
ImportError: numpy.core.multiarray failed to import
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
Segmentation fault

**EDIT 2:**
I fixed this error after reinstalling numpy. The version of numpy should be <1.24 otherwise you will get an error:

AttributeError: module 'numpy' has no attribute 'object'

Could you please provide some guidance on implementation object detection using SSD MobileNet V2 from scratch. The Face Detection example provided does not offer enough information to achieve this. Are there any other examples or tutorials that demonstrate how to implement object detection using the AI Deck?

**Comment by hmllr** [comment]
Hi :)
Yes, this should be possible - however, you'll have to put some effort into it. I think a good starting point will be the greenwaves mobilenet example: https://github.com/GreenWaves-Technologies/mobilenet/blob/public/README.md I haven't tried it myself, but as this is the same processor as on the AI-deck it should work. However, it is using images read from files, meaning you'll need to adapt it to read camera images (as in the face detection example or the classifier example). You'll most likely also need to train on augmented images or a custom dataset., as the himax camera gives you not very high quality images. Actually, I think the classifier example (https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/ai-examples/classification-demo) is closer to what you are looking for than the face detection example.
And lastly, here a paper that might be interesting (the code is not in the repo yet, this is just a preprint): https://arxiv.org/abs/2301.12175
Hope this helps,
Hanna

**Reply by ArtemRedchych** [reply]
Hello,
I am not very good in C and don't have experience in creating a Makefiles. Can i use Makefile from classification-demo if i just replace tflite file with object detection model?

**Reply by ArtemRedchych** [reply]
I have already put a lot of afford and still i have no progres. I have tryed to modify pre-trained [ssd_mobilenet_v2_320x320_coco17_tpu-8](http://download.tensorflow.org/models/object_detection/tf2/20200711/ssd_mobilenet_v2_320x320_coco17_tpu-8.tar.gz) model from tensorflow [garden](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md). The model works just fine in my jupiter notebook. I have exported to new_hope_converted_model.tflite and also quatitazed version off the model to new_hope_converted_model_q.tflite.

I am using a gap_sdk in docker image with autotiler **and the same makefile as in classification example(i have replaced classification_q.tflite with  my model).** When i try to build it i get this error during autotiler step:

Graph Node S132_Op_CONV_2D_0_49_trans_out0, Stacked tensor S245_Output replaces source binding S132_Output in Out
   Symbol:          S251_Output[   In] Adding   Edge From     S137_Op_CONV_2D_0_51_trans_out0 To                S253_Op_CUSTOM_0_101 New
   Symbol:          S251_Output[   In] Adding   Edge From     S185_Op_CONV_2D_0_69_trans_out0 To                S253_Op_CUSTOM_0_101 New
   
   .
   .
   .
      Symbol:             S9_Infos[   In] Adding   Edge From                      __GraphEntry__ To                 S9_Conv2d_16x32x1x1 New
   Sqemu: uncaught target signal 11 (Segmentation fault) - core dumped
make: *** [model_rules.mk:74: BUILD_MODEL_SQ8BIT/balloon_detectionKernels.c] Segmentation fault
make: *** [model_rules.mk:74: BUILD_MODEL_SQ8BIT/balloon_detectionKernels.c] Segmentation fault
 

I also have tried a model [lite-model_ssd_mobilenet_v1_100_320_uint8_default_1.tflite](https://tfhub.dev/iree/lite-model/ssd_mobilenet_v1_100_320/uint8/default/1) and i am getting an error in build process:

allocating 2 buffers of size 73728 Bytes
Reorganizing constant tensor for Arg Filter of Kernel S175_Conv2d_128x64x3x3_Relu6, Initial size: 73728 Bytes (73728 Bytes), Reorganized Size: 73728 Bytes
Moving Tile 0 [BYTE] from 0, LinesxLength = 128x144 -> Size 18432, Stride: 576 (576 2D) ItemSize: 9 ------> To 0, LinesxLength = 1x18432 -> Size 18432 Bits
Moving Tile 1 [BYTE] from 144, LinesxLength = 128x144 -> Size 18432, Stride: 576 (576 2D) ItemSize: 9 ------> To 18432, LinesxLength = 1x18432 -> Size 18432 Bits
Moving Tile 2 [BYTE] from 288, LinesxLength = 128x144 -> Size 18432, Stride: 576 (576 2D) ItemSize: 9 ------> To 36864, LinesxLength = 1x18432 -> Size 18432 Bits
Moving Tile 3 [BYTE] from 432, LinesxLength = 128x144 -> Size 18432, Stride: 576 (576 2D) ItemSize: 9 ------> To 55296, LinesxLength = 1x18432 -> Size 18432 Bits
Reorganization done
Loading coefficient file ./BUILD_MODEL_SQ8BIT/tensors/Featureextractormobilenetv1con.tensor: 128 Word items
Loading coefficient file ./BUILD_MODEL_SQ8BIT/tensors/S175_Mul_scale.tensor: 128 Byte items
Loading coefficient file ./BUILD_MODEL_SQ8BIT/tensors/S175_Mul_shift.tensor: 128 Byte items
Loading coefficient file ./BUILD_MODEL_SQ8BIT/tensors/S175_Infos.tensor: 9 Byte items
Loading coefficient file ./BUILD_MODEL_SQ8BIT/tensors/Boxpredictor_5boxencodingpredi_fd500e4b.tensor: 16384 Byte items
Error: For coefficient file ./BUILD_MODEL_SQ8BIT/tensors/Boxpredictor_5boxencodingpredi_fd500e4b.tensor expecting 16384 Byte items, got 128
Execution aborted
    CC classification.c
classification.c:32:10: fatal error: classificationKernels.h: No such file or directory
 #include "classificationKernels.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [/gap_sdk/rtos/freeRTOS/vendors/gwt/rules/freeRTOS_rules.mk:379: /module/examples/ai/classification/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/classification.o] Error 1


Can you please tell me what I am doing wrong? I need to make it work for my bachelor's thesis.

**Comment by hmllr** [comment]
Hi Artem,
Using the exact same Makefile will unfortunately not work, as the Makefile specifies which files should be compiled (here: https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/ai/classification/Makefile#L62). This means in classification.c it will look for the classificationKernels.h file which would have been generated for the classification example while fine-tuning/quantizing/exporting the tflite model to c code (https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/ai-examples/classification-demo/#fine-tune-a-pre-trained-image-classification-cnn). As this file doesn't exist it fails. You need to replace the files which should be compiled with your own files. Then you'll have a chance it'll work with almost the same Makefile - assuming your memory requirements are similar.
Hope this helps,
Hanna

**Reply by ArtemRedchych** [reply]
I changed application name so as all files with "classification" to "balloon_detection". As i see  model_decl.mk is responsible for creation ofballoon_detectionKernels.c and balloon_detectionKernels.h

```
MODEL_STATE = $(MODEL_BUILD)/$(MODEL_PREFIX).json
MODEL_SRC = $(MODEL_PREFIX)Model.c
MODEL_HEADER = $(MODEL_PREFIX)Info.h
MODEL_GEN = $(MODEL_BUILD)/$(MODEL_PREFIX)Kernels 
MODEL_GEN_C = $(addsuffix .c, $(MODEL_GEN))
MODEL_GEN_CLEAN = $(MODEL_GEN_C) $(addsuffix .h, $(MODEL_GEN))
MODEL_GEN_EXE = $(MODEL_BUILD)/GenTile
```

when i use custom model i am getting this error:

   Symbol: Ssd_mobile_net_v2_keras_featur_d73ac52d[   In] Adding   Edge From                      __GraphEntry__ To          S12_Conv2d_96x16x1x1_Relu6 Exists
   Symbol: Ssd_mobile_net_v2_keras_featur_0d00071f[   In] Adding   Edge From                      __GraphEntry__ To          S12_Conv2d_96x16x1x1_Relu6 Exists
   Symbol:             S9_Infos[   In] Adding   Edge From                 qemu: uncaught target signal 11 (Segmentation fault) - core dumped

when i use model lite-model_ssd_mobilenet_v1_100_320_uint8_nms_1.tflite from tensorflow hub i am getting this error:

make: *** No rule to make target 'BUILD_MODEL_SQ8BIT/balloon_detectionKernels.h', needed by '/module/examples/ai/balloon_detection/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/BUILD_MODEL_SQ8BIT/balloon_detectionKernels.o'.  Stop.

or

Error: For coefficient file ./BUILD_MODEL_SQ8BIT/tensors/Boxpredictor_5boxencodingpredi_ca3851d7.tensor expecting 16384 Byte items, got 128
Execution aborted
    CC balloon_detection.c
    CC com.c
    CC cpx.c
    CC balloon_detectionKernels.c
BUILD_MODEL_SQ8BIT/balloon_detectionKernels.c:1:10: fatal error: balloon_detectionKernels.h: No such file or directory
 #include "balloon_detectionKernels.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [/gap_sdk/rtos/freeRTOS/vendors/gwt/rules/freeRTOS_rules.mk:379: /module/examples/ai/balloon_detection/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/BUILD_MODEL_SQ8BIT/balloon_detectionKernels.o] Error 1


when i use body detection model from [gap8 example](https://github.com/GreenWaves-Technologies/body_detection) everything wokrs just fine. In each case i use the same makefile

**Reply by hmllr** [reply]
I am unfortunately not sure about this - it seems to be a problem with qemu though, and not with the Crazyflie. It seems to be a common error when using M1 processors, so if that's what you are using I'd try it on a different machine. Otherwise I would try to up/downgrade qemu to find a compatible version, but not sure if that's really the issue.
